### PR TITLE
Added `-CaseInsensitive` switch parameter to `Select-Object` & `Get-Unique` cmdlets

### LINKED ISSUE #12059: Select-Object -unique is inconsistently case sensitive
# Steps to reproduce

```powershell
"derp","Derp" | Select-Object -unique
```

# Expected behavior

```none
Single Return of derp
```

# Actual behavior

```none
[PSCustomObject]@{Name = "derp"},[PSCustomObject]@{Name = "Derp"} | Select-Object -unique
# Returns one 'derp'

"derp","Derp" | Select-Object -unique
# Returns two, "Derp", and "derp"

"derp","Derp" | Sort-Object -unique
# Returns One 'derp'

# Also: This is also the way it behaves in 5.1, if it helps
```

# Environment data

```none
Name                           Value
----                           -----
PSVersion                      7.0.0
PSEdition                      Core
GitCommitId                    7.0.0
OS                             Microsoft Windows 10.0.18362
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### PR DESCRIPTION
**Pull request by ArmaanMcleod**
# PR Summary

Fixes #12059

Added `-CaseInsensitive` switch parameter to `Select-Object` & `Get-Unique` cmdlets.

## PR Context

The `Select-Object` and `Get-Unique` cmdlets currently do case-sensitive string comparison when filtering unique items, and don't allow the user to override this behaviour and use case-insensitive comparison.

To make sure no breaking change is introduced, the most ideal solution is to introduce a `-CaseInsensitive` parameter.

#### Select-Object -Unique

```powershell
> "derp", "Derp" | Select-Object -Unique
derp
Derp

> "derp", "Derp" | Select-Object -Unique -CaseInsensitive
derp
```

#### Get-Unique

```powershell
> "aa", "Aa", "bb", "Bb" | Get-Unique
aa
Aa
bb
Bb

> "aa", "Aa", "bb", "Bb" | Get-Unique -AsString -CaseInsensitive
aa
bb

> "aa", "Aa", "bb", "Bb" | Get-Unique -CaseInsensitive
aa
bb
```

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: https://github.com/MicrosoftDocs/PowerShell-Docs/issues/10097
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetUnique.cs [hunk 1]
```diff
@@ -77,7 +84,7 @@ protected override void ProcessRecord()
                 if (string.Equals(
                     inputString,
                     _lastObjectAsString,
-                    StringComparison.CurrentCulture))
+                    CaseInsensitive.IsPresent ? StringComparison.CurrentCultureIgnoreCase : StringComparison.CurrentCulture))
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetUnique.cs:87]: Should we consider `OnType` above too?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/Select-Object.cs [hunk 2]
```diff
@@ -632,7 +639,7 @@ private void FilteredWriteObject(PSObject obj, List<PSNoteProperty> addedNotePro
                 bool isObjUnique = true;
                 foreach (UniquePSObjectHelper uniqueObj in _uniques)
                 {
-                    ObjectCommandComparer comparer = new(true, CultureInfo.CurrentCulture, true);
+                    ObjectCommandComparer comparer = new(true, CultureInfo.CurrentCulture, !CaseInsensitive.IsPresent);
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/Select-Object.cs:642]: Please use named argument for `true`.

Comment (ArmaanMcleod) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/Select-Object.cs]: Added named arguments. Included for `ascending` and `caseSensitive` for better clarity.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetUnique.cs (hunk 1 above)
Comment (ArmaanMcleod) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetUnique.cs:87]: @iSazonov `Ontype` would only return one instance of each type so case sensitivity shouldn't matter here?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetUnique.cs [hunk 3]
```diff
@@ -89,9 +96,9 @@ protected override void ProcessRecord()
             else // compare as objects
             {
                 _comparer ??= new ObjectCommandComparer(
-                    true, // ascending (doesn't matter)
+                    ascending: true, // ascending (doesn't matter)
                     CultureInfo.CurrentCulture,
-                    true); // case-sensitive
+                    caseSensitive: !CaseInsensitive.IsPresent); // case-sensitive
```
Comment (ArmaanMcleod) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetUnique.cs:101]: Given these use named arguments now maybe we can remove the comments as well?

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetUnique.cs]: Yes, please remove the old comments.

**Comment by SteveL-MSFT** [comment]
@PowerShell/wg-powershell-cmdlets reviewed this, we agree that adding a `-CaseInsensitive` switch makes sense particularly the internals also support it and just hard coded to be case sensitive.

**Review by iSazonov — APPROVED** [review]

**Review by SteveL-MSFT — APPROVED** [review]
LGTM

**Pull request merged**